### PR TITLE
Feature/cocktail sort

### PR DESCRIPTION
**Pull request by KianMiridoozini**
_feature/cocktail-sort : 
Comments Fixed
Sticky Fixed
Sorting Fixed
Likes Fixed

Comment functionalities have been fixed and adjusted.
Users can now add comments, then edit or delete the comments they have created. Users can also post and delete their reply to other comments.
These functionalities have been adjusted to refrain from refreshing the totality of the given page. 
This adjustment has been made through the separation of comment section from the original cocktails/views.php

No known issues regarding the comments.

However, profile.js, cocktail.js, admin.js, and user-management.js? require some adjustments 
### DO NOT ACCEPT THESE FILES IF THERE EXISITS A FUNCTIONING VERSION. 

Javascript had causd some conflicts where some functionalities stopped working after the implementation of sort and category js.

Notably:
- Likes (Fixed)
- Sorting (Fixed)
- Sticky images (Fixed)
- Comments (Fixed)
- User management ???
- Cocktail.js 
- Admin.js_

### **Update:**
feature/cocktail-sort: Major Fixed

Previous modules are now bundled inside a large main.js file inside modules folder.

Initialization of the modules have been converted into conditionals that depend on the page name. 
Page name variable has been added to some files. They are then retrieved from a meta tag inside the head.

Suggested changes are addressed.

Working initializations from the main.js:
- Likes
- Sort and category
- Sticky
- Random
- Profile
- Comments
- Cocktail edit
- Search

Admin functions are set on pause and are currently commented out as Brilliant Beaver has made some changes.

Known issue:
Exhaustion of events on home page. All functions related to homepage work, however if there is and overload of (specifically) sorting and category, the initialization may cause looping.

This shall be a simple fix for later iteration.

### CODE REVIEW

File: app/controllers/CommentController.php [hunk 1]
```diff
@@ -2,23 +2,24 @@
 class CommentController
 {
     private $commentService;
+    private $commentRepository;
     private $cocktailService;
 
-    public function __construct(CommentService $commentService, CocktailService $cocktailService)
+    public function __construct(CommentService $commentService, CommentRepository $commentRepository, CocktailService $cocktailService)
```
Comment (jayb2302) [inline app/controllers/CommentController.php]: I was thinking that we should adjust how dependencies are handled here to better align with the MVC + Repository pattern we’re following. Specifically, we should only inject the CommentService into the controller, instead of both the service and the repository. 

The controller’s role is to handle HTTP requests and responses. Any business logic or database access should be managed by the service and repository layers.
By relying only on the CommentService, the controller stays focused and avoids handling logic directly tied to the database.

File: app/controllers/CommentController.php [hunk 2]
```diff
@@ -27,109 +28,152 @@ public function addComment($cocktailId)
...
+            // Add the comment
+            $this->commentService->addComment($userId, $cocktailId, $commentText, $parentCommentId);
+
+            // Fetch updated cocktail and comments
+            $cocktail = $this->cocktailService->getCocktailById($cocktailId);
+            $comments = $this->commentService->getCommentsWithReplies($cocktailId);
+            $currentUser = (new AuthController())->getCurrentUser(); // Ensure current user is available
+
+            // Render the entire comment section
+            ob_start();
```
Comment (jayb2302) [inline app/controllers/CommentController.php:44]: Using ob_start in a controller tightly couples the controller with the view layer, which goes against the principle of separation of concerns. This makes testing the controller harder because rendering the view requires the full environment, including the filesystem.

File: app/controllers/CommentController.php [hunk 3]
```diff
@@ -27,109 +28,152 @@ public function addComment($cocktailId)
...
+            require __DIR__ . '/../views/cocktails/comment_section.php';
+            $commentsHtml = ob_get_clean();
+
+            // Return JSON response
+            header('Content-Type: application/json');
+            echo json_encode(['success' => true, 'html' => $commentsHtml]);
+        } catch (Exception $e) {
+            http_response_code(500); // Internal Server Error
+            echo json_encode(['error' => $e->getMessage()]);
```
Comment (jayb2302) [inline app/controllers/CommentController.php]: Returning raw error messages, for example ($e->getMessage()) in JSON responses can expose sensitive details about our application. This is a security risk.

Comment (jayb2302) [reply app/controllers/CommentController.php]: better approach would be 
```php
catch (Exception $e) {
    // Log the error internally and return a generic message
    error_log($e->getMessage());
    http_response_code(500); // Internal Server Error
    echo json_encode(['error' => 'An unexpected error occurred.']);
}

File: app/models/Cocktail.php [hunk 4]
```diff
@@ -3,7 +3,7 @@ class Cocktail
 {
     private $cocktail_id;
     private $user_id;
-    public $hasLiked = false;
+    public $hasLiked = null;
```
Comment (jayb2302) [inline app/models/Cocktail.php]: I noticed $hasLiked is set to null now instead of false. Is there a specific reason for this? Usually, it’s either true or false to track whether the cocktail is liked, so having null might add some unnecessary complexity unless we need it to mean something like ‘unknown’ or ‘not fetched yet. What's the reason for this change ?

File: app/repositories/CommentRepository.php [hunk 5]
```diff
@@ -45,27 +51,26 @@ public function getTopLevelCommentsByCocktailId($cocktailId)
...
+        foreach ($results as $result) { // Single loop only
+            $comments[] = new Comment(
+                $result['comment_id'],
+                $result['user_id'],
+                $result['username'],
+                $result['cocktail_id'],
+                $result['parent_comment_id'],
+                $result['comment'], // Ensure this matches your database
+                $result['created_at']
+            );
```
Comment (jayb2302) [inline app/repositories/CommentRepository.php:71]: Good call

File: routes.php [hunk 6]
```diff
@@ -97,6 +97,7 @@
 // Comment interactions
 $router->add('POST', '#^/cocktails/(\d+)-[^/]+/comments$#', [CommentController::class, 'addComment']);
 $router->add('GET', '#^/comments/(\d+)/edit$#', [CommentController::class, 'edit']); // Edit comment
+$router->add('POST', '#^/comments/(\d+)/edit$#', [CommentController::class, 'edit']); // Edit comment
```
Comment (jayb2302) [inline routes.php]: Wouldn’t it be clearer to use edit for the GET method and update for the POST method? It better reflects the purpose of each action and keeps the routes more RESTful 🛌💤😛

File: app/controllers/HomeController.php [hunk 7]
```diff
@@ -113,9 +113,7 @@ public function index($categoryName = null, $sortOption = 'recent')
         $isAdding = $action === 'add';
         $isLoggingIn = $action === 'login';
         $isRegistering = $action === 'register';
-        $includeScripts = [
-            asset('assets/js/sort-category.js')
-        ];  
+
```
Comment (jayb2302) [inline app/controllers/HomeController.php]: Can we keep this and ajust the path. I think we’ll need it for the AjaxController.

File: app/models/User.php [hunk 8]
```diff
@@ -106,6 +106,11 @@ public function canEditCocktail($cocktailUserId)
     {
         return !$this->isSuspended() && ($this->getId() === $cocktailUserId || $this->isAdmin());
     }
+    public function canEditComment($commentUserId)
+    {
+        return !$this->isSuspended() && ($this->getId() === $commentUserId || $this->isAdmin());
+    }
+
```
Comment (jayb2302) [inline app/models/User.php:113]: Nice

File: app/views/layout/footer.php [hunk 9]
```diff
@@ -5,18 +5,28 @@
...
+
+    if (document.querySelector('.category-sidebar')) initializeSortAndCategories();
+    if (document.querySelector('.like-button')) initializeLikes();
+    if (document.querySelector('#userTableBody')) initializeAdmin();
+    if (document.querySelector('.aboutContainer')) initializeRandomCocktail();
+    if (document.querySelector('.commentsSection')) initializeComments();
+    if (document.getElementById('edit-profile-form')) initializeProfile();
+    if (document.querySelector('#searchInput')) initializeSearch();
+    if (document.querySelector('.set-sticky')) initializeSticky();
+    if (document.getElementById('cocktailId')) initializeCocktail();
```
Comment (jayb2302) [inline app/views/layout/footer.php]: Can you combine the modules into a main.js file and just call main.js in the footer instead?🙏🏽

**Review by jayb2302 — CHANGES_REQUESTED** [review]

File: app/models/Cocktail.php (hunk 4 above)
Comment (KianMiridoozini) [reply app/models/Cocktail.php]: Upon refreshing a page, even the cocktails that had been liked by the user, would display with a white heart. Only after reapplying like, the heart color would change to red.
My deduction lead to this change. Of course if your system does not reflect the same issue as mine, I would need to reinvestigate why this happens.

File: routes.php (hunk 6 above)
Comment (KianMiridoozini) [reply routes.php]: Whilst that is true, the method used in the CommentController does not rely or call for a GET,  because of how the JS is retrieving data.
This meant that I needed to decide on the uri extension on whether I keep edit or update.
Since we call it as to edit a comment and not to update a comment, the dessicion on my end was to use edit for the POST.

Of course given the changes I need to make to the controller this might change, and we can keep edit for GET and POST for update.

File: app/controllers/HomeController.php (hunk 7 above)
Comment (KianMiridoozini) [reply app/controllers/HomeController.php]: This was the cause of one of the errors on the console. I did not investigate deeply about it, as removing this line, removed the error whilst keeping the functionality the same with no compromise.

File: app/views/layout/footer.php (hunk 9 above)
Comment (KianMiridoozini) [reply app/views/layout/footer.php]: I can combine the modules into a main.js file, and investigate how I can dynamically call for initializeXXXXX in a more optimized manner.
Currently, the reason for the "if" statements are to call for a module, only when a certain element selector or element id is called. By using these "if" statements, i am currently ensuring that there would be no overlap or overwrite of the JS modules regardless of the loaded page.

File: app/controllers/CommentController.php (hunk 1 above)
Comment (KianMiridoozini) [reply app/controllers/CommentController.php]: You are absolutely right. I shall make some adjustments to this.

File: app/controllers/CommentController.php (hunk 2 above)
Comment (KianMiridoozini) [reply app/controllers/CommentController.php:44]: I shall investigate how to achieve the desired function without calling for ob_start

File: app/controllers/CommentController.php (hunk 3 above)
Comment (KianMiridoozini) [reply app/controllers/CommentController.php]: Yes this was an oversight from my behalf, whilst i was cleaning logs for debugging. 
I have commented out most of the messages and debugging lines, but this one went under my radar.

File: app/controllers/HomeController.php (hunk 7 above)
Comment (jayb2302) [reply app/controllers/HomeController.php]: yes but can you comment it out

File: app/views/layout/footer.php (hunk 9 above)
Comment (jayb2302) [reply app/views/layout/footer.php]: the logic still remains the same its just about keeping footer cleaner

File: app/controllers/BaseController.php [hunk 10]
```diff
@@ -8,11 +8,13 @@
 require_once __DIR__ . '/../repositories/DifficultyRepository.php';
 require_once __DIR__ . '/../repositories/LikeRepository.php';
 require_once __DIR__ . '/../repositories/UnitRepository.php';
+require_once __DIR__ . '/../repositories/CommentRepository.php';
```
Comment (jayb2302) [inline app/controllers/BaseController.php:11]: no comment

**Review by jayb2302 — APPROVED** [review]
Nice!
